ui/src/plugins/dev.perfetto.ExplorePage/query_builder/sub_query_node.ts

```diff
@@ -0,0 +1,95 @@
+// Copyright (C) 2025 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import m from 'mithril';
+import protos from '../../../protos';
+import {
+  createFinalColumns,
+  createSelectColumnsProto,
+  nextNodeId,
+  NodeType,
+  QueryNode,
+  QueryNodeState,
+} from '../query_node';
+import {
+  createFiltersProto,
+  createGroupByProto,
+} from './operations/operation_component';
+import {ColumnInfo, newColumnInfoList} from './column_info';
+
+export class SubQueryNode implements QueryNode {
+  readonly nodeId: string;
+  readonly type = NodeType.kSubQuery;
+  readonly prevNode?: QueryNode;
+  readonly nextNode?: QueryNode;
+  readonly sourceCols: ColumnInfo[];
+  readonly finalCols: ColumnInfo[];
+  readonly state: QueryNodeState;
+
+  constructor(state: QueryNodeState) {
+    this.nodeId = nextNodeId();
+    this.state = state;
+    this.prevNode = state.prevNode;
+    this.sourceCols = this.prevNode!.finalCols;
```

Comment:
I understand that `prevNode` being undefined is a developer error, but we usually use `assertExists(this.prevNode)` instead of `!` so we know about it immediately.